docs/architecture.md

```diff
@@ -0,0 +1,345 @@
+```rust
+impl fuel_poa_consensus::ports::BlockProducer for Service<BlockProducer> {
+    // insert impl here
+}
+```
+
+### Block Importer
+
+#### Ports: fuel_block_importer::ports
+
+```rust
+trait Relayer {
+    // wait until the relayer is synced with ethereum
+    async fn await_synced() -> Result<()>;
+    // wait until a specific da_height is reached
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+}
+
+trait Executor {
+    async fn validate_and_store_block(block: &FuelBlock) -> Result<()>;
+}
+
+trait TransactionPool {
+    // remove a set of txs from the pool after a block is committed
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Database {
+    // insert consensus information associated with a sealed block
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+```
+#### Adapters: fuel_core::service::adapters::block_importer
+```rust
+impl fuel_poa_consensus::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+```
+
+### Synchronizer
+
+#### Ports: fuel_sync::ports
+```rust
+pub struct Entity<Data> {
+    pub data: Data,
+    pub sender: PeerId,
+};
+
+trait PeerToPeer {
+    async fn fetch_best_network_block_header() -> Result<Entity<SealedFuelBlockHeader>>;
+    async fn fetch_blocks(query: Range<BlockHeight>) -> Result<Entity<Vec<SealedFuelBlock>>>;
+    async fn report_invalid_height(peer: PeerId) -> Result<()>;
+    async fn report_invalid_block(peer: PeerId) -> Result<()>;
+}
+
+trait BlockImporter {
+    // commit a sealed block to the chain
+    async fn commit(block: SealedFuelBlock) -> Result<()>;
+}
+
+trait Consensus {
+    // check with the consensus layer whether a block header passes consensus rules
+    async fn validate_sealed_block_header(block: SealedFuelBlockHeader) -> Result<()>;
+}
+
+trait Database {
+    // get current fuel blockchain height
+    fn get_current_height() -> Result<BlockHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::fuel_sync
+_N/A_
+
+### P2P
+
+#### Ports: fuel_p2p::ports
+```rust
+trait Database {
+    // used when other peers are syncing
+    fn get_sealed_block(height: BlockHeight) -> Result<SealedFuelBlock>; 
+}
+```
+
+#### Adapters: fuel_core::service::adapters::peer_to_peer
+```rust
+impl fuel_poa_consensus::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_txpool::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+```
+
+### Transaction Pool
+
+#### Ports: fuel_txpool::ports
+```rust
+trait PeerToPeer {
+    async fn broadcast_transaction(transaction: Transaction) -> Result<()>;
+    fn report_invalid_tx(peer_id: PeerId) -> Result<()>;
+    // handler for gossipped transactions
+    fn on_new_tx(tx: Transaction, peer: PeerId);
```

Comment:
```suggestion
    fn received_network_transactions() -> impl Stream<(Transaction, PeerId)>;
```

Reply:
In the traits, we can't return `impl`, so it will be `Box<dyn Stream<...>>`.

Also, `Entity` from the suggestion for FS can be a global type, and we can use it here `Stream<Entity<Transaction>>`=)

Reply:
And another question, do we want to return `Stream` or use `async next_new_tx() -> Result<...>`?

Because we need to use the same pattern in all services=)

Reply:
Good call on still needing `BoxStream` (wen GAT?). 

`async next_data_item() -> Result<..>` is also a good option and less complicated. We can just leave it up to each domain-specific adapter to figure out the details of how to reliably provide this data to the domain (e.g. cloning tokio pubsub receivers). This way, the adapter won't be forced to always create a stream if the underlying data source is some other kind of custom future.

The main reason to stick with a stream in the domain interface is that it's the most conventional way to represent an async future that will yield multiple values. However, from a consumer perspective this just comes down to having a `stream.next().await` instead of a `future.await`. Having our own async `port.next_something().await` methods seems fairly similar semantically to calling next on a stream and erases some of the pains of dealing with streams directly.